File: leetcode/98.validate-binary-search-tree.py
# @leet start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from math import inf


class Solution:
    def isValidBST(self, root: TreeNode | None) -> bool:
        return self.helper(root, -inf, inf)

    def helper(self, node: TreeNode | None, min_: int, max_: int) -> bool:
        if not node:
            return True

        if node.val <= min_ or node.val >= max_:
            return False

        return self.helper(node.left, min_, min(max_, node.val)) and self.helper(
            node.right, max(min_, node.val), max_
        )


# @leet end

File: leetcode/98.validate-binary-search-tree.ts
// @leet start
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function isValidBST(root: TreeNode | null): boolean {
  return helper(root, -Infinity, Infinity);
}

function helper(root: TreeNode | null, min: number, max: number): boolean {
  return (
    !root ||
    (root.val > min &&
      root.val < max &&
      helper(root.left, min, root.val) &&
      helper(root.right, root.val, max))
  );
}
// @leet end
